RecoCTPPS/TotemRPLocal/plugins/TotemRPLocalTrackFitter.cc

```diff
@@ -0,0 +1,185 @@
+*   Jan Kašpar (jan.kaspar@gmail.com)
+*
+****************************************************************************/
+
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/ESWatcher.h"
+
+#include "DataFormats/Common/interface/DetSetVector.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPRecHit.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPUVPattern.h"
+#include "DataFormats/CTPPSReco/interface/TotemRPLocalTrack.h"
+
+#include "Geometry/Records/interface/VeryForwardRealGeometryRecord.h"
+#include "Geometry/VeryForwardGeometryBuilder/interface/TotemRPGeometry.h"
+
+#include "RecoCTPPS/TotemRPLocal/interface/TotemRPLocalTrackFitterAlgorithm.h"
+
+//----------------------------------------------------------------------------------------------------
+
+/**
+ *\brief Fits tracks trough a single RP.
+ **/
+class TotemRPLocalTrackFitter : public edm::stream::EDProducer<>
+{
+  public:
+    explicit TotemRPLocalTrackFitter(const edm::ParameterSet& conf);
+
+    virtual ~TotemRPLocalTrackFitter() {}
+
+    virtual void produce(edm::Event& e, const edm::EventSetup& c);
+
+  private:
+    int verbosity_;
+
+    /// Selection of the pattern-recognition module.
+    edm::InputTag tagUVPattern;
+
+    edm::EDGetTokenT<edm::DetSetVector<TotemRPUVPattern>> patternCollectionToken;
+    
+    /// A watcher to detect geometry changes.
+    edm::ESWatcher<VeryForwardRealGeometryRecord> geometryWatcher;
+    
+    /// The instance of the fitter module
+    TotemRPLocalTrackFitterAlgorithm fitter_;
+};
+
+//----------------------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------------------
+
+using namespace std;
+using namespace edm;
+
+//----------------------------------------------------------------------------------------------------
+
+TotemRPLocalTrackFitter::TotemRPLocalTrackFitter(const edm::ParameterSet& conf)
+   : verbosity_(conf.getParameter<int>("verbosity")), fitter_(conf)
+{
+  tagUVPattern = conf.getParameter<edm::InputTag>("tagUVPattern");
+  patternCollectionToken = consumes<DetSetVector<TotemRPUVPattern>>(tagUVPattern);
+
+  produces<DetSetVector<TotemRPLocalTrack>>();
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void TotemRPLocalTrackFitter::produce(edm::Event& e, const edm::EventSetup& setup)
+{
+  if (verbosity_ > 5)
+    printf(">> TotemRPLocalTrackFitter::produce\n");
+
+  // get geometry
+  edm::ESHandle<TotemRPGeometry> geometry;
+  setup.get<VeryForwardRealGeometryRecord>().get(geometry);
+
+  if (geometryWatcher.check(setup))
```

Comment:
Actually, I believe this use of a watcher is correct. The two lines above are doing the data request every event which is what we advise. However, the watcher is used to decide when the `fitter_` should flush its cache and recalcuate the information it needs from the geometry. As I commented on the declaration of `Reset` I think it would be clearer if `Reset` took the geometry as an argument so that it would update the cache in `Reset` instead of deferring till `FitTrack`.